# Rapport du Devoir 1, IFT3275-A

Auteurs:
-    Ding Jun, QIU
-    Mahamadou

In [1]:
import decimal as d
import numpy as np

# Question 1

Le code se trouve dans py

In [2]:
# Clé publique Question 1.1
q1_N = 143516336909281815529104150147210248002789712761086900059705342103220782674046289232082435789563283739805745579873432846680889870107881916428241419520831648173912486431640350000860973935300056089286158737579357805977019329557985454934146282550582942463631245697702998511180787007029139561933433550242693047924440388550983498690080764882934101834908025314861468726253425554334760146923530403924523372477686668752567287060201407464630943218236132423772636675182977585707596016011556917504759131444160240252733282969534092869685338931241204785750519748505439039801119762049796085719106591562217115679236583
q1_e = 3

# Cryptogramme 1.1
q1_C = 1101510739796100601351050380607502904616643795400781908795311659278941419415375

In [3]:
import devoir1
question1 = devoir1.codeBreaker_RSA()

msg = question1.Decipher(q1_N,q1_e, q1_C)
print("The deciphered M is: "+str(msg))
print("The UTF-8 encoding of the message is: "+question1.int_to_str(msg))

using N-th root attack!
The deciphered M is: 103275247704828660521722734
The UTF-8 encoding of the message is: Umberto Ecn


In [4]:
print("what the actual M should be without data imprecision: "+question1.int_to_str(msg+1))

what the actual M should be without data imprecision: Umberto Eco


The actual original string should probably be "Umberto Eco" which gives a M of 103275247704828660521722735 (which is our M+1)

This difference is probably due to our nth-root attack implementation uses C**(1/e) to calculate n-th root. If there is a way to calculate n-th root without the division then there would probably be a lot less precision loss.

And also, if there is no division, we wouldn't have to transform integer values into Decimal (because these int are too big for a division resulting in float, so we need to use Decimal division operations)

In [5]:
# A form of meet-in-middle (in name only) exhaustive attack is also implemented to break RSA that don't satisfy the n-th root attack condition but still short enough
# I've set it to allow exhaustive attack on RSA with bit length shorter than 51 

# why 51? because I was brute-forcing Q1.2 with it and it can run through 1mil options in under 10 seconds...  of course as it gets longer the time cost will increase linearly
# limit being it can only calculate small value enough to be stored in np uint64 (so the upper bit length limit is 64).
# Python Int can be way longer and in that case we must use np array dtype=object, but in that case the speed would be at least 100 times slower
# although, if we really do exhaustive attack on 64 bit, it would take at least a whole day... in the scope of this homework I deem this unreasonable time.
# So I limit it to only do exhaustive attack if N bit length is less than 51 (worst expected runtime about 10 minutes?)

# using the RSA example given in class handouts
print("N=33 E=13 C=26, M should be 5? "+str(question1.Decipher(33, 13, 26)))

N length is reasonable for an exhaustive attack
Expected runtime: 0.0 minutes
intermediate result is: 26
N=33 E=13 C=26, M should be 5? 5


In [6]:
# Clé publique Question 1.2
q2_N = 172219604291138178634924980176652297603347655313304280071646410523864939208855547078498922947475940487766894695848119416017067844129458299713889703424997977808694983717968420001033168722360067307143390485095229367172423195469582545920975539060699530956357494837243598213416944408434967474317474605697904676813343577310719430442085422937057220239881971046349315235043163226355302567726074269720408051461805113819456513196492192727498270702594217800502904761235711809203123842506621973488494670663483187137290546241477681096402483981619592515049062514180404818608764516997842633077157249806627735448350463
q2_e = 173

# Cryptogramme 1.2
q2_C = 25782248377669919648522417068734999301629843637773352461224686415010617355125387994732992745416621651531340476546870510355165303752005023118034265203513423674356501046415839977013701924329378846764632894673783199644549307465659236628983151796254371046814548224159604302737470578495440769408253954186605567492864292071545926487199114612586510433943420051864924177673243381681206265372333749354089535394870714730204499162577825526329944896454450322256563485123081116679246715959621569603725379746870623049834475932535184196208270713675357873579469122917915887954980541308199688932248258654715380981800909

print("The deciphered M is: "+str(question1.Decipher(q2_N,q2_e, q2_C)))

cannot perform attack on such RSA within reasonable timeframe.
The deciphered M is: OPERATION ABORTED


I cannot find known weaknesses when e is bigger than 3, and with this length it's pratically impossible to break using exhaustive methods.

I've seen that some suggest Coppersmith method, and it does seem to break Q1.1 efficiently (tested on CoCalc, return the accurate result with no data loss, but CoCalc does not allow calculating Q1.2 due to long calculation time). But I can't get Sagemath to work locally on my setup, as such I cannot test its performance on Q1.2 and I can't tell whether Coppersmith's performance on Q1.1 is proof of it being a good method, or it was due Q1.1 being too weak.

...and also, with a N of 2001 bit-length I doubt any conventional method can break it in a reasonable timeframe... (especially when I re-read the class handout and was reminded that the world record for RSA break was only on a 800 bit length... RSA is good for a reason, and those people are way smarter and they probably used everything they have. If even them cannot handle such N, I doubt I can do it with my limited knowledge)

But! although the length of C and N is beyond my capabilities, the length of M (therefore, the possible range values of M) is much much smaller, and we can probably brute-force it. 

Like in the case of "Umberto Eco", its UTF-8 int form has a bit-length less than 90, and it has a possiblity range of 2**(11*8) = 309,485,009,821,345,068,724,781,056 = 256**11 (the UTF-8 form of it)

But, since we know the plaintext is a name, and a name does not use all possible range of a UTF-8 encoding (we dont use characters such as + - ( ) 1 2 3 4 in a name, and we only use capital letter at the beginning of a name), and so the possible range of a name is possibily millions time smaller than computing integer possibilities.

I was about to write a name generator to test that out, but this is where I re-read the question, and it specified "auteur célèbre", which is again a range of possibility much much smaller than the possible range of all names.

If it's famous, then it probably has a Wikipedia page. And if its in the category of 'auteur', then there is probably a category page on Wikipedia. And I was right.

So, I went on Wikipedia and copied all their entry names in the author category :

from 
https://en.wikipedia.org/wiki/List_of_authors_by_name:_A
to
https://en.wikipedia.org/wiki/List_of_authors_by_name:_Z

(the list of names can be found in authors.csv)

In [7]:
# These 2 functions are copied from Devoir 1 Examples

# convert string to list of integer
def str_to_int_list(x):
  z = [ord(a) for a in x  ]
  for x in z:
    if x > 256:
      #print(x)
      return False
  return z

# convert a strint to an integer
def str_to_int(x):
  x = str_to_int_list(x)
  if x == False:
    #print("Le text n'est pas compatible!")
    return False

  res = 0
  for a in x:
    res = res * 256 + a
  i = 0
  res = ""
  for a in x:
    ci = "{:08b}".format(a )
    #if len(ci)>8:
      #print()
      #print("long",a)
      #print()
    res = res + ci
  res = eval("0b"+res)
  return res

In [8]:
import csv

# Here, the list of author names are stored and parsed. The latter part in parenthesis are getting removed
with open('authors.csv', newline='', encoding='utf8') as csvfile:
    data = list(csv.reader(csvfile, delimiter='(')) # remove what comes after "("

# Now, we iterate through all author names, compute the corresponding C, and compare with the question.
for index in range(len(data)):
    name = data[index][0]
    n = name[:len(name) - 1]    # cut out the last ' ' in name entry
    i = str_to_int(n)

    c = i**q2_e % q2_N
    if (c == q2_C): # here, we get a match, so we print the author name and stop
        print("The ciphered author name is: "+n)
        #print(c)
        break

The ciphered author name is: Marcel Proust


This wouldn't be possible if the "Auteur" info was not available. 

Writing a name generator that would iterate through possible name combinations would be also quite tedious...

# Question 2



In [9]:
# On commence par obtenir un statistique sur le frequence des bi-characteres

bi_array = Q2.cut_string_into_pairs(corpus[10000:10500])
bi_freq = Q2.frequencyAnalysis(bi_array)
print(bi_freq[0:150,])


NameError: name 'Q2' is not defined

In [ ]:
# Then we compute the frequency of single characters
mono_freq = Q2.frequencyAnalysis(list(corpus[10000:10500]))
#print(mono_freq)


In [ ]:
merged_freq = Q2.mergeFreq(mono_freq, bi_freq)

(167, 2)
[' ' '0.154']
0.154
0
[[' ' '0.154']
 ['e' '0.112']
 ['s' '0.08']
 ['r' '0.074']
 ['t' '0.062']
 ['n' '0.054']
 ['i' '0.048']
 ['a' '0.048']
 ['u' '0.046']
 ['l' '0.044']
 ['o' '0.04']
 [' l' '0.028']
 ['es' '0.028']
 ['e ' '0.028']
 ['le' '0.028']
 ['ns' '0.024']
 ['é' '0.024']
 ['c' '0.022']
 ['p' '0.022']
 ['\r\n' '0.02']
 ['re' '0.02']
 [' p' '0.02']
 ['d' '0.02']
 ['nt' '0.02']
 ['s ' '0.02']
 [' t' '0.02']
 ['eu' '0.02']
 ['la' '0.016']
 [' s' '0.016']
 ['it' '0.016']
 ['ra' '0.016']
 [' c' '0.016']
 [',' '0.016']
 ['v' '0.016']
 ['ur' '0.016']
 ['t ' '0.016']
 ['ér' '0.016']
 ['s,' '0.016']
 ['m' '0.014']
 ['\r' '0.014']
 ['\n' '0.014']
 ['on' '0.012']
 ['oi' '0.012']
 ['te' '0.012']
 ['r ' '0.012']
 [', ' '0.012']
 ['rs' '0.012']
 [' v' '0.012']
 ['ti' '0.012']
 ['da' '0.012']]


In [ ]:
cypherText = "11010101011011000111010011001111000011100110110011001111110011000000110111001010010111010111111011001100011011001100111101101100011000011100111100001110011011001100111111001100000011011100101001011101011111101100110001101100011101001100111100001110011011001100111100111111000100100000011000111101000100100110110011001111011011000110000110010101000011000000111001101100110011110010000011001110011000010110110001110100110011110000111001101100110011110010000011001110011000010110110011001111011011000110000111001111000011100110110011001111001111110001001000000110001111010001001001101100011111001100111101011010001011011100101011001111011011001101010111001111"
cypherArray = Q2.cutCypherText(cypherText)
print(cypherArray)
cypherFreq = Q2.frequencyAnalysis(cypherArray)


charcount 82
[213 108 116 207  14 108 207 204  13 202  93 126 204 108 207 108  97 207
  14 108 207 204  13 202  93 126 204 108 116 207  14 108 207  63  18   6
  61  18 108 207 108  97 149  12  14 108 207  32 206  97 108 116 207  14
 108 207  32 206  97 108 207 108  97 207  14 108 207  63  18   6  61  18
 108 124 207  90  45 202 207 108 213 207]


In [ ]:
# The previous merged keys is the actual key table used, but the frequency calculated here does not reflect the real frequency
# for a given char c at index i (str[i]), if it and its previous (str[i-1:i+1]) is a key then c will not be counted in frequency

In [ ]:
def replaceByFrequency(cypherArray, cypherFreq, stdFreq, tryFirst = 50):
        
    string = []
    for x in range(len(cypherArray)):
        index = np.where(cypherFreq[:,0] == cypherArray[x])[0][0]
        #print(index)
        if (index < stdFreq.shape[0]) : string.append(str(stdFreq[index][0]))
        else: string.append(str("?"))

    print(string)

replaceByFrequency(cypherArray, cypherFreq, merged_freq)

original = ['e', 's', ',', ' ', 'le', 's', ' ', 'n', 'a', 't', 'i', 'o', 'n', 's', ' ', 's', 'ur', ' ', 'le', 's', ' ', 'n', 'a', 't', 'i', 'o', 'n', 's', ',', ' ', 'le', 's', ' ', 'v', 'é', 'r', 'it', 'é', 's', ' ', 's', 'ur', '\r', '\n', 'le', 's', ' ', 'er', 're', 'ur', 's', ',', ' ', 'le', 's', ' ', 'er', 're', 'ur', 's', ' ', 's', 'ur', ' ', 'le', 's', ' ', 'v', 'é', 'r', 'it', 'é', 's', '.', ' ', 'T', 'ou', 't', ' ', 's', 'e', ' ']
print(original)

['u', 'e', 'i', ' ', 's', 'e', ' ', 'n', 'ns', 'a', ' l', 'e ', 'n', 'e', ' ', 'e', 'r', ' ', 's', 'e', ' ', 'n', 'ns', 'a', ' l', 'e ', 'n', 'e', 'i', ' ', 's', 'e', ' ', 'es', 't', 'o', 'é', 't', 'e', ' ', 'e', 'r', 'c', ' p', 's', 'e', ' ', 'le', 'l', 'r', 'e', 'i', ' ', 's', 'e', ' ', 'le', 'l', 'r', 'e', ' ', 'e', 'r', ' ', 's', 'e', ' ', 'es', 't', 'o', 'é', 't', 'e', '\r\n', ' ', 'p', 're', 'a', ' ', 'e', 'u', ' ']
['e', 's', ',', ' ', 'le', 's', ' ', 'n', 'a', 't', 'i', 'o', 'n', 's', ' ', 's', 'ur', ' ', 'le', 's', ' ', 'n', 'a', 't', 'i', 'o', 'n', 's', ',', ' ', 'le', 's', ' ', 'v', 'é', 'r', 'it', 'é', 's', ' ', 's', 'ur', '\r', '\n', 'le', 's', ' ', 'er', 're', 'ur', 's', ',', ' ', 'le', 's', ' ', 'er', 're', 'ur', 's', ' ', 's', 'ur', ' ', 'le', 's', ' ', 'v', 'é', 'r', 'it', 'é', 's', '.', ' ', 'T', 'ou', 't', ' ', 's', 'e', ' ']


In [ ]:
# we can see that occurence test did not yield good result, only ' ' is a useful tool
# but here comes a question, 'es' is in the dictionary, so why "es" is being separately encoded as 'e' and 's' ?
# if the algorithm really prioritize pair, then we should see :
# 'es' ', ' 'le' 's ' 'n' 'at' 'io' 'ns' ' s' 'ur' (...) there's already a big difference

In [ ]:
# repurpose the function provided in question

def M_vers_symboles(M, K):
    encoded_text = []
    i = 0

    while i < len(M):
        # Vérifie les paires de caractères
        if i + 1 < len(M):
            pair = M[i] + M[i + 1]
            #print(pair)
            if pair in K:
                encoded_text.append(pair)
                i += 2  # Sauter les deux caractères utilisés
                continue

        # Vérifie le caractère seul
        #if M[i] in K:
        #    encoded_text.append(M[i])
        #else:
            # Conserve le caractère tel quel si non trouvé
        encoded_text.append(M[i])
        i += 1

    return encoded_text

def make_dict(keys):

    dictionary = {}
    for m in keys:
        dictionary[str(m)]="{:08b}".format(0)
    return dictionary

dct = make_dict(merged_freq[:,0])
print(dct)
text = corpus[10000:10100]
print(M_vers_symboles(text, dct))

{' ': '00000000', 'e': '00000000', 's': '00000000', 'r': '00000000', 't': '00000000', 'n': '00000000', 'i': '00000000', 'a': '00000000', 'u': '00000000', 'l': '00000000', 'o': '00000000', ' l': '00000000', 'es': '00000000', 'e ': '00000000', 'le': '00000000', 'ns': '00000000', 'é': '00000000', 'c': '00000000', 'p': '00000000', '\r\n': '00000000', 're': '00000000', ' p': '00000000', 'd': '00000000', 'nt': '00000000', 's ': '00000000', ' t': '00000000', 'eu': '00000000', 'la': '00000000', ' s': '00000000', 'it': '00000000', 'ra': '00000000', ' c': '00000000', ',': '00000000', 'v': '00000000', 'ur': '00000000', 't ': '00000000', 'ér': '00000000', 's,': '00000000', 'm': '00000000', '\r': '00000000', '\n': '00000000', 'on': '00000000', 'oi': '00000000', 'te': '00000000', 'r ': '00000000', ', ': '00000000', 'rs': '00000000', ' v': '00000000', 'ti': '00000000', 'da': '00000000', 'h': '00000000', '. ': '00000000', 'rr': '00000000', 'ue': '00000000', 've': '00000000', "'e": '00000000', 'sa': '0

we can see that, as I suspect, if we follow the logic then something is wrong.

I suspect the actual dictionary used to produce the cryptogram is not the one shown.

meaning, we cannot rely solely on frequency, I suspect we need to analyze french syntax and calculate/deduce the most likely combinations

- we know ' ' is reliable predictor (for now), and anything that is between 2 spaces and is short enough might be computed to deduce 'les' 'des' 'de' 'a' so on.
- but we dont actually know if ' ' is good predictor ? this might just be a lucky coincidence. if a specific dictionary has a bunch of char+' ' or ' '+char then we might not even see occurence of ' ' and mistake ' ' for 'e' 

- one way to check might be to get the highest frequency words and compare with our sample. we can see in merged_freq that ' ' and 'e' all have frequencies double than the 3rd place 's', so when we analyse the cyphertext, if we do not see 2 entry with frequencies distinguished above the rest, then we could assume at least one of them is being hidden.

- if one is hidden, then we have 2 cases that we could test concurrently.
- if both are hidden, then... we might have to calculate how much symbols are required to do so, and if a lot of symbols are used to hide ' ' and 'e' then the rest might be exposed such as 's' 

In [ ]:
# Now, back to cyphertext analysis

for x in range(cypherFreq.shape[0]):
    print(str(int(cypherFreq[x][0])) + "    "+ str( cypherFreq[x][1]) + "   "+str(np.where(cypherArray == cypherFreq[x][0])))

207    0.2073170731707317   (array([ 3,  6, 14, 17, 20, 29, 32, 39, 46, 52, 55, 60, 63, 66, 74, 78, 81]),)
108    0.2073170731707317   (array([ 1,  5, 13, 15, 19, 27, 31, 38, 40, 45, 50, 54, 59, 61, 65, 72, 79]),)
14    0.07317073170731707   (array([ 4, 18, 30, 44, 53, 64]),)
97    0.06097560975609756   (array([16, 41, 49, 58, 62]),)
18    0.04878048780487805   (array([34, 37, 68, 71]),)
204    0.04878048780487805   (array([ 7, 12, 21, 26]),)
116    0.036585365853658534   (array([ 2, 28, 51]),)
202    0.036585365853658534   (array([ 9, 23, 77]),)
213    0.024390243902439025   (array([ 0, 80]),)
206    0.024390243902439025   (array([48, 57]),)
6    0.024390243902439025   (array([35, 69]),)
93    0.024390243902439025   (array([10, 24]),)
63    0.024390243902439025   (array([33, 67]),)
126    0.024390243902439025   (array([11, 25]),)
32    0.024390243902439025   (array([47, 56]),)
13    0.024390243902439025   (array([ 8, 22]),)
61    0.024390243902439025   (array([36, 70]),)
149    0.0121

We can see 207 and 108 has frequency way above the rest. We can safely assume 207 and 108 are ' ' and 'e', though which is which might still be debatable.

Maybe there's a way to distinguish 'e' and ' ' ?

Although I cannot be sure, but knowing french, its highly unlikely that np.int64(108) is ' ' due to fact it repeat often in interval of 1.

and since 108 is unlikely to be ' ', the other option for ' ' is np.int64(207)

on constate que ce dict sur-represente les 's', et evite de faire apparaitre des elements repetitifs (erreur -> 'er' 're' 'ur' 's')
sachant les positions de ' ', y a t-il moyen de distinguer 's' et 'e' ?

les noms pluriels en francais suivent cette regle: char[]+'s '+char[]+'s '.

Sachant les positions du ' ', on peut utiliser cette regle pour distinguer 's' du 'e'

In [ ]:
import requests
import student_code as Q2


url = "https://www.gutenberg.org/ebooks/13846.txt.utf-8"  # Example URL (replace with your desired URL)
corpus = Q2.load_text_from_web(url)
url = "https://www.gutenberg.org/ebooks/4650.txt.utf-8"  # Example URL (replace with your desired URL)
corpus = corpus + Q2.load_text_from_web(url)

print(corpus[10000:10500])

es, les nations sur les nations, les vérités sur
les erreurs, les erreurs sur les vérités. Tout se tient dans l'univers;
mais qui pourrait tracer la ligne? On peut du moins entrevoir ce rapport
général; on peut dire que, sans cette foule d'erreurs qui ont inondé le
monde, Descartes peut-être n'eût point trouvé la route de la vérité.
Ainsi chaque philosophe en s'égarant avançoit le terme. Mais, laissant
là les temps trop reculés, je veux chercher dans le siècle même de
Descartes, ou dans c


In [ ]:
result_3 = []
result_4 = []
result_5 = []
result_6 = []

analysis = Q2.ngrams_count(corpus,3)
counter = 0
# only check 500 results. cuz latter half minimal return
for x in analysis:
    counter += 1
    result_3.append(Q2.checkFrequency(corpus, ''.join(x[0]),x[1], len(x[0]), True))
    if (counter > 500): break

analysis = Q2.ngrams_count(corpus,4)
counter = 0
# only check 500 results. cuz latter half minimal return
for x in analysis:
    counter += 1
    result_4.append(Q2.checkFrequency(corpus, ''.join(x[0]),x[1], len(x[0]), True))
    if (counter > 500): break

analysis = Q2.ngrams_count(corpus,5)
counter = 0
# only check 500 results. cuz latter half minimal return
for x in analysis:
    counter += 1
    result_5.append(Q2.checkFrequency(corpus, ''.join(x[0]),x[1], len(x[0]), True))
    if (counter > 500): break

analysis = Q2.ngrams_count(corpus,6)
counter = 0
# only check 500 results. cuz latter half minimal return
for x in analysis:
    counter += 1
    result_6.append(Q2.checkFrequency(corpus, ''.join(x[0]),x[1], len(x[0]), True))
    if (counter > 500): break

symbol [es ] frequency [1.037%] mean interval [95.40123520329388] minVal [3] maxVal [2940] median [39.0]
symbol [ qu] frequency [1.017%] mean interval [95.30260394792104] minVal [3] maxVal [20889] median [63.0]
symbol [ de] frequency [0.924%] mean interval [107.04009706494107] minVal [2] maxVal [3011] median [70.0]
symbol [que] frequency [0.769%] mean interval [127.47301234910503] minVal [3] maxVal [12213] median [73.0]
symbol [de ] frequency [0.726%] mean interval [136.62983669265853] minVal [2] maxVal [3138] median [87.0]
symbol [nt ] frequency [0.701%] mean interval [141.55492914825538] minVal [4] maxVal [4546] median [86.0]
symbol [ent] frequency [0.693%] mean interval [143.0873113643363] minVal [2] maxVal [3139] median [88.0]
symbol [ le] frequency [0.551%] mean interval [179.8205426356589] minVal [2] maxVal [9796] median [101.0]
symbol [ue ] frequency [0.532%] mean interval [186.19069580910366] minVal [3] maxVal [16289] median [104.0]
symbol [re ] frequency [0.488%] mean interval

In [ ]:
import csv

result = []

header = ['Symbol','Count','Frequency','MedianOccurence','MinOccurence']
with open('freq_ngram_6.csv', 'w', encoding='UTF8') as f:
    writer = csv.writer(f)
    writer.writerow(header)
    for x in result_6:
        #string = str(x[0])+','+str(x[1])+','+x[2]+x[3]+x[4]
        writer.writerow(x)
        print(x)


NameError: name 'result_6' is not defined

In [ ]:
import csv

stats_ngram_1 = []
stats_ngram_2 = []



# Here, the list of author names are stored and parsed. The latter part in parenthesis are getting removed
with open('freq_ngram_2.csv', encoding='utf8') as csvfile:
    data = list(csv.reader(csvfile)) # remove what comes after "("

    # Now, we iterate through all author names, compute the corresponding C, and compare with the question.
    for index in range(len(data)):
        if (index == 0): continue
        if (len(data[index]) < 1) :continue
        
        temp = [data[index][0],int(data[index][1]),float(data[index][2]),float(data[index][3]),int(data[index][4])]
        print(temp)
        stats_ngram_2.append(temp)

    

    

['e ', 37131, 3.9636883405797954, 24.227518784842854, 2]
['s ', 23196, 2.4761443200584132, 39.37808242800483, 2]
['t ', 19119, 2.0409296109327815, 47.99633872064439, 2]
['es', 17636, 1.8826211945400142, 52.10376502608301, 2]
[' d', 15967, 1.7044575081209121, 57.658545750610635, 2]
['\n\n', 15429, 1.6470266733135563, 59.71540605353555, 1]
['en', 14753, 1.574864509131823, 62.48742628617908, 1]
['qu', 13069, 1.3950995912589843, 70.55704338510981, 3]
[' l', 12572, 1.3420454557585089, 73.45943366210626, 2]
['re', 11827, 1.2625176268895866, 78.16081846622136, 1]
[' p', 11780, 1.2575004350012116, 78.51426146010186, 2]
['nt', 11725, 1.251629253004177, 78.84690831556503, 2]
['de', 11625, 1.2409543766459326, 79.56232258064516, 2]
['le', 11511, 1.2287850175975337, 80.37798627399879, 2]
['on', 11147, 1.1899284676535233, 83.02825872432045, 1]
[', ', 11100, 1.1849112757651485, 83.38747747747747, 2]
[' c', 10945, 1.1683652174098693, 84.55258108725445, 2]
[' e', 10298, 1.099298767372027, 89.9658186055

Currently at lead:

https://stackoverflow.com/questions/12488722/counting-bigrams-pair-of-two-words-in-a-file-using-python

In [1]:
# CheckRepeat On Corpus
import requests
import student_code as Q2

"""

url = "https://www.gutenberg.org/ebooks/13846.txt.utf-8"  # Example URL (replace with your desired URL)
corpus = Q2.load_text_from_web(url)
url = "https://www.gutenberg.org/ebooks/4650.txt.utf-8"  # Example URL (replace with your desired URL)
corpus = corpus + Q2.load_text_from_web(url)

print("Analyzing string repeat form")
print("-- Monogram 1 repeat")
print(Q2.checkRepeat(corpus,1,0,15))
print(Q2.checkRepeat(corpus,1,1,15))
print(Q2.checkRepeat(corpus,1,2,15))
print(Q2.checkRepeat(corpus,1,3,15))
print(Q2.checkRepeat(corpus,1,4,15))
print(Q2.checkRepeat(corpus,1,5,15))
print(Q2.checkRepeat(corpus,1,6,15))
print(Q2.checkRepeat(corpus,1,7,15))

print("-- Monogram 2 repeat")
print(Q2.checkRepeat(corpus,2,0,15))
print(Q2.checkRepeat(corpus,2,1,15))
print(Q2.checkRepeat(corpus,2,2,15))
print(Q2.checkRepeat(corpus,2,3,15))
print(Q2.checkRepeat(corpus,2,4,15))
print(Q2.checkRepeat(corpus,2,5,15))
print(Q2.checkRepeat(corpus,2,6,15))
print(Q2.checkRepeat(corpus,2,7,15))


print("-- Monogram 3 repeat")
print(Q2.checkRepeat(corpus,3,0,15))
print(Q2.checkRepeat(corpus,3,1,15))
print(Q2.checkRepeat(corpus,3,2,15))
print(Q2.checkRepeat(corpus,3,3,15))
print(Q2.checkRepeat(corpus,3,4,15))
print(Q2.checkRepeat(corpus,3,5,15))
print(Q2.checkRepeat(corpus,3,6,15))
print(Q2.checkRepeat(corpus,3,7,15))


print("-- Monogram 4 repeat")
print(Q2.checkRepeat(corpus,4,0,15))
print(Q2.checkRepeat(corpus,4,1,15))
print(Q2.checkRepeat(corpus,4,2,15))
print(Q2.checkRepeat(corpus,4,3,15))
print(Q2.checkRepeat(corpus,4,4,15))
print(Q2.checkRepeat(corpus,4,5,15))
print(Q2.checkRepeat(corpus,4,6,15))
print(Q2.checkRepeat(corpus,4,7,15))

print("-- Monogram 5 repeat")
print(Q2.checkRepeat(corpus,5,0,15))
print(Q2.checkRepeat(corpus,5,1,15))
print(Q2.checkRepeat(corpus,5,2,15))
print(Q2.checkRepeat(corpus,5,3,15))
print(Q2.checkRepeat(corpus,5,4,15))
print(Q2.checkRepeat(corpus,5,5,15))
print(Q2.checkRepeat(corpus,5,6,15))
print(Q2.checkRepeat(corpus,5,7,15))

print("-- Monogram 6 repeat")
print(Q2.checkRepeat(corpus,6,0,15))
print(Q2.checkRepeat(corpus,6,1,15))
print(Q2.checkRepeat(corpus,6,2,15))
print(Q2.checkRepeat(corpus,6,3,15))
print(Q2.checkRepeat(corpus,6,4,15))
print(Q2.checkRepeat(corpus,6,5,15))
print(Q2.checkRepeat(corpus,6,6,15))
print(Q2.checkRepeat(corpus,6,7,15))

SyntaxError: incomplete input (2607843493.py, line 5)

In [3]:
# WordCount test on Corpus
import requests
import student_code as Q2


url = "https://www.gutenberg.org/ebooks/13846.txt.utf-8"  # Example URL (replace with your desired URL)
corpus = Q2.load_text_from_web(url)
url = "https://www.gutenberg.org/ebooks/4650.txt.utf-8"  # Example URL (replace with your desired URL)
corpus = corpus + Q2.load_text_from_web(url)

In [4]:
corpus_wordCount = []

corpus_wordCount.append(Q2.wordCountTest(corpus,1,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,2,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,3,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,4,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,5,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,6,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,7,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,8,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,9,0))
corpus_wordCount.append(Q2.wordCountTest(corpus,10,0))
#print(Q2.wordCountTest(corpus,11,0,100))
#print(Q2.wordCountTest(corpus,12,0,100))

for x in corpus_wordCount:
    print(x[:30])

[[' ', np.int64(142842)], ['e', np.int64(112881)], ['s', np.int64(59875)], ['n', np.int64(52866)], ['t', np.int64(52105)], ['i', np.int64(51339)], ['u', np.int64(47497)], ['a', np.int64(45607)], ['r', np.int64(43845)], ['o', np.int64(42049)], ['l', np.int64(35149)], ['d', np.int64(25063)], ['c', np.int64(22926)], ['p', np.int64(20138)], ['m', np.int64(20120)], ['\r', np.int64(15429)], ['\n', np.int64(15428)], ['q', np.int64(13100)], ['é', np.int64(12539)], [',', np.int64(11981)], ['v', np.int64(9824)], ["'", np.int64(8958)], ['f', np.int64(7186)], ['h', np.int64(5836)], ['g', np.int64(5486)], ['b', np.int64(5176)], ['j', np.int64(4686)], ['.', np.int64(4591)], ['à', np.int64(2970)], ['x', np.int64(2736)]]
[['e ', np.int64(37131)], ['s ', np.int64(23196)], ['t ', np.int64(19119)], ['es', np.int64(17636)], [' d', np.int64(15967)], ['\r\n', np.int64(15428)], ['en', np.int64(14753)], ['qu', np.int64(13069)], [' l', np.int64(12572)], ['re', np.int64(11827)], [' p', np.int64(11780)], ['nt', 

In [4]:
for x in corpus_wordCount[0]:
    maxRepeat = 1
    repeat = True
    #print(x[0][0])
    while repeat:
      repeat = False
      matchStr = ''
      for i in range(maxRepeat+1):
         matchStr += x[0][0]
      #print(matchStr)
      for i in corpus_wordCount[maxRepeat]:
         if (i[0] == matchStr):
            maxRepeat += 1
            repeat = True
            break
      #if (repeat):   print("match in "+str(maxRepeat-1))
      if (maxRepeat > 5): repeat = False
    if (maxRepeat > 3) : print(str(x[0][0]) + " max repeat "+str(maxRepeat))

  max repeat 6
. max repeat 4
* max repeat 6


In [12]:
stats_ngram_1 = Q2.loadStatsNgram('freq_ngram_1.csv')
index = 1
for x in stats_ngram_1:
    index += 2
    repeat = Q2.checkConsecutiveChar(corpus, x[0],1,False)
    if repeat > 3 : print("symbol "+str(x[0])+" index "+str(index)+ " maxrepeat "+ str(repeat))

symbol   index 3 maxrepeat 16
symbol . index 57 maxrepeat 4
symbol * index 165 maxrepeat 38


In [16]:
'''
stats_ngram_2 = Q2.loadStatsNgram('freq_ngram_2.csv')
index = 1
for x in stats_ngram_2:
    index += 2
    repeat = Q2.checkConsecutiveChar(corpus, x[0],2,False)
    if repeat > 3 : print("symbol "+str(x[0])+" index "+str(index)+ " maxrepeat "+ str(repeat))
'''
print("__ maxConsec "+str(Q2.checkConsecutiveChar(corpus, '  ',2,False)) +" total "+ str(Q2.countWord(corpus, '  ',2)))
print("rn maxConsec "+str(Q2.checkConsecutiveChar(corpus, '\r\n',2,False)) +" total "+ str(Q2.countWord(corpus, '\r\n', 2)))
print("nr maxConsec "+str(Q2.checkConsecutiveChar(corpus, '\n\r',2,False)) +" total "+ str(Q2.countWord(corpus, '\n\r', 2)))

__ maxConsec 8 total 584
rn maxConsec 8 total 15428
nr maxConsec 7 total 1562


In [20]:
import random

a = random.randint(3400, 7200)
b = random.randint(36000, 65000)
l = a+b
c = random.randint(0, len(corpus)-l)

M = corpus[c:c+l]
# . is masked by ._
#print("._ maxConsec "+str(Q2.checkConsecutiveChar(M, '. ',2,False)) +" total "+ str(Q2.countWord(M, '. ', 2)))
print("* maxConsec "+str(Q2.checkConsecutiveChar(M, '*',1,False)) +" total "+ str(Q2.countWord(M, '*',1)))
#print(". maxConsec "+str(Q2.checkConsecutiveChar(M, '.',1,False)) +" total "+ str(Q2.countWord(M, '.',1)))
print("nr maxConsec "+str(Q2.checkConsecutiveChar(M, '\n\r',2,False)) +" total "+ str(Q2.countWord(M, '\n\r', 2)))
print("rn maxConsec "+str(Q2.checkConsecutiveChar(M, '\r\n',2,False)) +" total "+ str(Q2.countWord(M, '\r\n', 2)))
print("rn maxConsec "+str(Q2.checkConsecutiveChar(M, '. ',2,False)) +" total "+ str(Q2.countWord(M, '. ', 2)))

* maxConsec 0 total 0
nr maxConsec 4 total 101
rn maxConsec 5 total 696
rn maxConsec 1 total 190


In [2]:
print("nr maxConsec "+str(Q2.checkConsecutiveChar(corpus, '\n\r',2,False)) +" total "+ str(Q2.countWord(corpus, '\n\r', 2)))

print("rn maxConsec "+str(Q2.checkConsecutiveChar(corpus, '\r\n',2,False)) +" total "+ str(Q2.countWord(corpus, '\r\n', 2)))

'''
print("n maxConsec "+str(Q2.checkConsecutiveChar(corpus, '\n',1,False)) +" total "+ str(Q2.countWord(corpus, '\n', 1)))

print("r maxConsec "+str(Q2.checkConsecutiveChar(corpus, '\r',1,False)) +" total "+ str(Q2.countWord(corpus, '\r', 1)))

print("er maxConsec "+str(Q2.checkConsecutiveChar(corpus, 'e\r',2,False)) +" total "+ str(Q2.countWord(corpus, 'e\r', 2)))

print("sr maxConsec "+str(Q2.checkConsecutiveChar(corpus, 's\r',2,False)) +" total "+ str(Q2.countWord(corpus, 's\r', 2)))

print("tr maxConsec "+str(Q2.checkConsecutiveChar(corpus, 't\r',2,False)) +" total "+ str(Q2.countWord(corpus, 't\r', 2)))
'''


NameError: name 'Q2' is not defined

In [ ]:
# We now have a way to somewhat reliably find \r\n
# let's write an algorithm that counts all occurence before \r\n

import numpy as np

def countEOL(corpus, eolSymbol = '\r\n', eolSymbolLen = 2, countSymbol = -1):
    word = eolSymbol
    array = []
    wordlen = eolSymbolLen

    start = -countSymbol

    if start < 0: start = 0

    for x in range(start, len(corpus) - wordlen):
      if corpus[x:x+wordlen] == word:
        if countSymbol < 0: array.append(corpus[x+countSymbol:x])
        else: array.append(corpus[x+wordlen:x+wordlen+countSymbol])
    return array

def sortEOL(arr):
  sorted = []
  nparr = np.unique(arr, return_counts=True, return_index=True)
  sort = np.flip(nparr[2].argsort())
  for x in sort:
    arrayIndex = nparr[1][x]
    sorted.append([arr[arrayIndex],nparr[2][x]])
  return sorted

In [18]:
array = countEOL(corpus, eolSymbol='.', eolSymbolLen= 1, countSymbol=-1)
#print(array[:10])
sorted = sortEOL(array)
for x in sorted[:10]:
   print(x)

['s', np.int64(980)]
['t', np.int64(370)]
['r', np.int64(282)]
['n', np.int64(230)]
['é', np.int64(114)]
['i', np.int64(97)]
['1', np.int64(90)]
['E', np.int64(85)]
['.', np.int64(85)]
['u', np.int64(78)]


In [19]:
array = countEOL(corpus, eolSymbol='.\r\n', eolSymbolLen= 3, countSymbol=-2)
#print(array[:10])
sorted = sortEOL(array)
for x in sorted[:10]:
   print(x)

['re', np.int64(75)]
['nt', np.int64(63)]
['le', np.int64(57)]
[' B', np.int64(42)]
['se', np.int64(37)]
['er', np.int64(37)]
['de', np.int64(37)]
['on', np.int64(32)]
['te', np.int64(29)]
['té', np.int64(24)]


In [20]:
array = countEOL(corpus, eolSymbol='\r\n', eolSymbolLen= 2, countSymbol=2)
#print(array[:10])
sorted = sortEOL(array)
for x in sorted[:10]:
   print(x)

['qu', np.int64(694)]
['de', np.int64(513)]
['co', np.int64(492)]
['le', np.int64(320)]
['pa', np.int64(308)]
["l'", np.int64(270)]
['en', np.int64(246)]
['ce', np.int64(244)]
['pr', np.int64(239)]
['et', np.int64(227)]


In [24]:
array = countEOL(corpus, eolSymbol='quelqu', eolSymbolLen= 6, countSymbol=2)
#print(array[:20])
sorted = sortEOL(array)
for x in sorted[:10]:
   print(x)

['es', np.int64(165)]
['ef', np.int64(49)]
['e\r', np.int64(38)]
["'u", np.int64(34)]
['’u', np.int64(1)]


In [1]:
import test

t1 = test.TestDecryption()
t1.test_decryption_accuracy()


charcount 66954
38 max repeat 4 maximumConsecutive 4 total 1293
EOL rn FOUND : 38
pre-EOL period FOUND : 116
[['49', np.int64(2)], ['8', np.int64(1)]]
[[np.int64(219), np.int64(2)], [np.int64(10), np.int64(2)]]
post-EOL qu FOUND : 39
quelqu FOUND : 49
que_ FOUND : 219
ques FOUND : 10
esles FOUND : 64
esdes FOUND : -1
qui_est FOUND : 238
est FOUND : 26
PRINTING DICTIONARY ....



{'38': '\r\n', '116': '.', '39': 'qu', '49': 'el', '219': 'e ', '10': 'es', '64': ' l', '-1': ' d', '238': 'i ', '26': 't '}
Deciphering ....
Similarité : 0.31%
Différences entre les messages :
--- Original
+++ Déchiffré
@@ -1,1777 +1,1294 @@
-nceau de
-l'éloquence; et j'affoiblirois l'éloge d'un grand homme en cherchant à
-peindre ce qui ne doit être que calculé. Contentons-nous de remarquer
-ici que, par son analyse, Descartes fit faire plus de progrès à la
-géométrie qu'elle n'en avoit fait depuis la création du monde. Il
-abrégea les travaux, il multiplia les forces, il donna une nouvelle
-marche à l'esprit

AssertionError: 0.003124739605032914 not greater than or equal to 0.995 : La similarité est seulement de 0.31%, ce qui est inférieur à 99.5 %.